cirq-google/cirq_google/engine/abstract_engine.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+import datetime
+from typing import Dict, List, Optional, Sequence, Set, Union
+
+import cirq
+from cirq_google.engine import abstract_program, abstract_processor
+from cirq_google.engine.client import quantum
+from cirq_google.serialization import Serializer
+
+
+class AbstractEngine(ABC):
+    """An abstract object representing a collection of quantum procexsors.
+
+    Each processor within the AbstractEngine can be referenced by a string
+    identifier through the get_processor interface.
+
+    The Engine interface also includes convenience methods to access
+    programs, jobs, and sampler.
+
+    This is an abstract interface and inheritors must implement the abstract methods.
+
+    """
+
+    @abstractmethod
+    def get_program(self, program_id: str) -> abstract_program.AbstractProgram:
+        """Returns an exsiting AbstractProgram given an identifier.
+
+        Args:
+            program_id: Unique ID of the program within the parent project.
+
+        Returns:
+            An AbstractProgram object for the program.
+        """
+
+    @abstractmethod
+    def list_programs(
+        self,
+        created_before: Optional[Union[datetime.datetime, datetime.date]] = None,
+        created_after: Optional[Union[datetime.datetime, datetime.date]] = None,
+        has_labels: Optional[Dict[str, str]] = None,
+    ) -> List[abstract_program.AbstractProgram]:
+        """Returns a list of previously executed quantum programs.
+
+        Args:
+            created_after: retrieve programs that were created after this date
+                or time.
+            created_before: retrieve programs that were created after this date
+                or time.
+            has_labels: retrieve programs that have labels on them specified by
+                this dict. If the value is set to `*`, filters having the label
+                regardless of the label value will be filtered. For example, to
```

Comment:
Should "**filters** having the label" be "**programs** having the label"?

Also, minor nit: as a verb, "filter" is ambiguous (e.g. "filter out X" vs "filter down to X") - it's unclear if matches will be kept or removed. "...will be **returned.**" may be clearer.

Similarly elsewhere in this PR.